From 4f5773958cc95261395f1118bd5c656984e1345e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Mon, 20 Jul 2026 00:32:08 +0200
Subject: [PATCH] prefs: Add proper handler for delete-event

Connect the preferences window's delete-event signal to a proper
handler that hides the window and returns TRUE instead of using the
close button callback directly.

The close button callback has an incompatible signature for
delete-event and does not return a gboolean. As a result, closing the
window with the window manager close button could still destroy the
window, causing errors and an empty window when it was opened again.
---
 data/ui/prefs.glade | 2 +-
 src/gui/gui-prefs.c | 8 ++++++++
 src/gui/gui-prefs.h | 1 +
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/data/ui/prefs.glade b/data/ui/prefs.glade
index d6cc424a..e52210aa 100644
--- a/data/ui/prefs.glade
+++ b/data/ui/prefs.glade
@@ -129,7 +129,7 @@
     <property name="destroy-with-parent">True</property>
     <property name="type-hint">dialog</property>
     <property name="skip-taskbar-hint">True</property>
-    <signal name="delete-event" handler="on_prefs_close_clicked" swapped="no"/>
+    <signal name="delete-event" handler="on_prefs_delete_event" swapped="no"/>
     <child>
       <object class="GtkVBox" id="vbox1">
         <property name="visible">True</property>
diff --git a/src/gui/gui-prefs.c b/src/gui/gui-prefs.c
index 172b7088..3851e64f 100644
--- a/src/gui/gui-prefs.c
+++ b/src/gui/gui-prefs.c
@@ -611,6 +611,14 @@ void on_prefs_close_clicked (GtkWidget* widget, void* user) {
     config_save ();
 }
 
+G_MODULE_EXPORT
+gboolean
+on_prefs_delete_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
+{
+    on_prefs_close_clicked (widget, user_data);
+    return TRUE;
+}
+
 G_MODULE_EXPORT
 void on_prefs_reset_clicked (GtkWidget* widget, void* user) {
     config_load_defaults ();
diff --git a/src/gui/gui-prefs.h b/src/gui/gui-prefs.h
index fc06db6a..0ee0a85b 100644
--- a/src/gui/gui-prefs.h
+++ b/src/gui/gui-prefs.h
@@ -88,6 +88,7 @@ void toggle_compilestatus (GtkWidget* widget, void* user);
 void toggle_spaces_instof_tabs (GtkWidget* widget, void* user);
 void toggle_autosaving (GtkWidget* widget, void* user);
 void on_prefs_close_clicked (GtkWidget* widget, void* user);
+gboolean on_prefs_delete_event (GtkWidget *widget, GdkEvent *event, gpointer user_data);
 void on_prefs_reset_clicked (GtkWidget* widget, void* user);
 void on_tabwidth_value_changed (GtkWidget* widget, void* user);
 void on_configure_snippets_clicked (GtkWidget* widget, void* user);
